File: Problems/No.2195_Append_K_Integers_With_Minimal_Sum/append_k_integers_with_minimal_sum.py
class Solution(object):
    def minimalKSum(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: int
        """

        # parameters
        appended_sum = 0
        added_num = 0
        i = 1

        # start computing
        nums.sort()
        nums.insert(0, 0)
        
        while (k > 0):

            # the case between two nums
            added_num = nums[i] - nums[i-1] - 1
            if (added_num > 0):
                added_num = min(added_num, k)
                appended_sum += ((nums[i-1]+1) + (nums[i-1]+added_num))*added_num//2
                k -= added_num
            i += 1

            # the case after all nums
            if ((i >= len(nums)) and (k > 0)):
                appended_sum += ((nums[-1]+1) + (nums[-1]+k))*k//2
                break

        return appended_sum

File: Problems/No.2195_Append_K_Integers_With_Minimal_Sum/append_k_integers_with_minimal_sum.cpp
class Solution {
public:
    long long minimalKSum(vector<int>& nums, int k) {
        
        // parameter
        long long appended_sum = 0;
        int added_num = 0;
        int i = 1;

        // start computing
        sort(nums.begin(), nums.end());
        nums.insert(nums.begin(), 0);
        
        while (k > 0) {

            // the case between two nums
            added_num = nums[i] - nums[i-1] - 1;
            if (added_num > 0) {
                
                added_num = std::min(added_num, k);
                appended_sum += ((nums[i-1]+1) + (nums[i-1]+added_num))*(long)added_num/2;
                k -= added_num;

            }
            i++;

            // the case after all nums
            if ((i >= nums.size()) && (k > 0)) {
                
                appended_sum += ((nums[nums.size()-1]+1) + (nums[nums.size()-1]+k))*(long)k/2;
                break;

            }

        }
        
        return appended_sum;

    }
};
